src/karapace/schema_reader.py

```diff
@@ -433,9 +443,10 @@ def consume_messages(self, msgs: list[Message], watch_offsets: bool) -> None:
             # Default keymode is CANONICAL and preferred unless any data consumed
             # has key in non-canonical format. If keymode is set to DEPRECATED_KARAPACE
             # the subsequent keys are omitted from detection.
-            if not self.ready and self.key_formatter.get_keymode() == KeyMode.CANONICAL:
-                if msg_keymode == KeyMode.DEPRECATED_KARAPACE:
-                    self.key_formatter.set_keymode(KeyMode.DEPRECATED_KARAPACE)
+            with self._ready_lock:
+                if not self._ready and self.key_formatter.get_keymode() == KeyMode.CANONICAL:
```

Comment:
With the changes done the key mode detection could be run during the normal operation. Not anymore only at startup.